File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashSet;
import java.util.Set;
import java.util.StringTokenizer;


public class Main {
	static int N, max;
	static int[] dp, array;
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		dp = new int[N+1];
		array = new int[N+1];

		StringTokenizer st = new StringTokenizer(br.readLine());
		
		for(int n = 1; n <= N; n++) array[n] = Integer.parseInt(st.nextToken());

		for(int i = 1; i <= N; i++){
			dp[i] = 1;
			for(int j = 1; j <= i; j++){
				if(array[i] > array[j]) dp[i] = Math.max(dp[i], dp[j]+1);
			}
			max = Math.max(max, dp[i]);
		}
		System.out.println(max);
	}

}

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun main(){
    val br = BufferedReader(InputStreamReader(System.`in`))
    val N = br.readLine().toInt()
    val numbers : IntArray = br.readLine().split(" ").map { it.toInt() }.toIntArray()
    var dp  = IntArray(N+1)
    var result = 0
    // 인덱스를 기준으로 처음부터 자기자신까지 오는 수 중 자기보다 작은 수를 비교
    for(i in 1 .. N) {
        // 자기자신은 증가하는 부분에 포함함
        dp[i] = 1
        for(j in 1 .. i){
            if(numbers[i-1] <= numbers[j-1]) continue;
            dp[i] = maxOf(dp[i],dp[j]+1)
        }
        result = maxOf(result, dp[i])
    }
    println(result)
}